### PR TITLE
cmake-ng integration

### PR DESCRIPTION
**Pull request by dotnwat**
builds the demos using smf via `add_subdirectory` which will also recursively build the dependencies for smf like seastar...

usage

```
mkdir build
cd build
cmake ..
```

TODO:

also want integration via the `-config.cmake` method so users can _share_ a single smf build tree.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "smf"]
```
Comment (emaxerrno) [inline .gitmodules]: nice.

**Comment by emaxerrno** [comment]
Thinking about this, and the purpose of this project is to have a more real world sample project. Should we do the cmakelists.txt.in that we do in upstream smf?

**Comment by dotnwat** [comment]
@senior7515 that's a good idea, but that will require changes to the smf install configuration so that it correctly points to the dependencies that are built as part of the smf build.

**Comment by emaxerrno** [comment]
@noahdesu  did you manage to finish this branch #4  needs the fix.

**Comment by dotnwat** [comment]
I can put up a fix for the integration style that uses `add_subdirectory(smf)` tonight. The integration that also builds all the deps like is done in smf at configuration time is going to take some more hacking on the installation side of smf, unless you've hacked on that a bit in the past couple weeks?

**Comment by emaxerrno** [comment]
@noahdesu  - i think i fixed the smf install targets and it works. It is largely a copy of the cmakelist.txt.in from the smf repo + adding smf as a ExternalProject_Add(). 

However, I think the add_subdirectory() is probably more desirable for *us* - the maintainers. 

I would prefer that. 

Thoughts?

**Comment by dotnwat** [comment]
updated. works with latest smf master

**Comment by dotnwat** [comment]
well, travis might not work but it worked locally.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -1,23 +1,19 @@
-find_package(smf)
-include(smfc_generator)
+set(SMF_MANAGE_DEPS ON CACHE "" INTERNAL FORCE)
+set(SMF_BUILD_PROGRAMS OFF CACHE "" INTERNAL FORCE)
+add_subdirectory(smf EXCLUDE_FROM_ALL)
```
Comment (emaxerrno) [inline src/CMakeLists.txt:3]: nice.

**Comment by emaxerrno** [comment]
@noahdesu can you remove cooking.sh & recipe/*

File: .travis.yml [hunk 3]
```diff
@@ -4,7 +4,7 @@ language: cpp
...
 script:
   - sudo ./install-deps.sh
-  - ./travis_stdout.sh "./cooking.sh -r wellknown -t Release -- -DCMAKE_CXX_FLAGS_RELEASE='-O1 -DNDEBUG' > ./travis_cmake.log"
-  - tail --lines=500 ./travis_cmake.log
-  - ninja -C build
-  - echo "FINISHED! SUCCESS!"
+  - mkdir build
+  - cd build
+  - cmake ..
+  - make
```
Comment (emaxerrno) [inline .travis.yml:10]: `make -j$(nproc)` ?

File: README.md [hunk 4]
```diff
@@ -9,11 +9,19 @@ getting started for smf - The Fastest RPC In The West
...
-1. ./build.sh
-2. cd build/release
-3. Run the apps
-    *   ./src/demo_server -c 1 & (`run on one core; background`)
-    *   ./src/demo_client -c 1   (`run on one core`)
+Build the demo server and client, which will also build smf and its
+dependencies. **Note that the initial build may take a long time**.
+
+```bash
+cmake .
```
Comment (emaxerrno) [inline README.md:16]: we should make it one line ppl can copy paste

`mkdir -p build && cmake .. && make`

Comment (emaxerrno) [reply README.md:16]: `make -j$(nproc)`

**Comment by emaxerrno** [comment]
@noahdesu This is great. Just some minor cosmetic fixes on the doc & on the travis.yml

**Pull request merged**